pyccl/halos/halo_model.py

```diff
@@ -0,0 +1,557 @@
+        return i1 + self.mbf0 * array_2[..., 0]
+
+    def profile_norm(self, cosmo, a, prof):
+        """ Returns :math:`I^0_1(k\\rightarrow0,a|u)`
+        (see :meth:`~HMCalculator.I_0_1`).
+
+        Args:
+            cosmo (:class:`~pyccl.core.Cosmology`): a Cosmology object.
+            a (float): scale factor.
+            prof (:class:`~pyccl.halos.profiles.HaloProfile`): halo
+                profile.
+
+        Returns:
+            float or array_like: integral value.
+        """
+        # Compute mass function
+        self._get_ingredients(a, cosmo, False)
+        uk0 = prof.fourier(cosmo, self.precision['k_min'],
+                           self.mass, a, mass_def=self.mass_def).T
+        norm = 1. / self._integrate_over_mf(uk0)
+        return norm
+
+    def I_0_1(self, cosmo, k, a, prof):
+        """ Solves the integral:
+
+        .. math::
+            I^0_1(k,a|u) = \\int dM\\,n(M,a)\\,\\langle u(k,a|M)\\rangle,
+
+        where :math:`n(M,a)` is the halo mass function, and
+        :math:`\\langle u(k,a|M)\\rangle` is the halo profile as a
+        function of scale, scale factor and halo mass.
+
+        Args:
+            cosmo (:class:`~pyccl.core.Cosmology`): a Cosmology object.
+            k (float or array_like): comoving wavenumber in Mpc^-1.
+            a (float): scale factor.
+            prof (:class:`~pyccl.halos.profiles.HaloProfile`): halo
+                profile.
+
+        Returns:
+            float or array_like: integral values evaluated at each
+            value of `k`.
+        """
+        # Compute mass function
+        self._get_ingredients(a, cosmo, False)
+        uk = prof.fourier(cosmo, k, self.mass, a,
+                          mass_def=self.mass_def).T
+        i01 = self._integrate_over_mf(uk)
+        return i01
+
+    def I_1_1(self, cosmo, k, a, prof):
+        """ Solves the integral:
+
+        .. math::
+            I^1_1(k,a|u) = \\int dM\\,n(M,a)\\,b(M,a)\\,
+            \\langle u(k,a|M)\\rangle,
+
+        where :math:`n(M,a)` is the halo mass function,
+        :math:`b(M,a)` is the halo bias, and
+        :math:`\\langle u(k,a|M)\\rangle` is the halo profile as a
+        function of scale, scale factor and halo mass.
+
+        Args:
+            cosmo (:class:`~pyccl.core.Cosmology`): a Cosmology object.
+            k (float or array_like): comoving wavenumber in Mpc^-1.
+            a (float): scale factor.
+            prof (:class:`~pyccl.halos.profiles.HaloProfile`): halo
+                profile.
+
+        Returns:
+            float or array_like: integral values evaluated at each
+            value of `k`.
+        """
+        # Compute mass function and halo bias
+        self._get_ingredients(a, cosmo, True)
+        uk = prof.fourier(cosmo, k, self.mass, a,
+                          mass_def=self.mass_def).T
+        i11 = self._integrate_over_mbf(uk)
+        return i11
+
+    def I_0_2(self, cosmo, k, a, prof_2pt, prof1, prof2):
```

Comment:
I think you need None defaults here. I'd put prof1 first, and then default prof_2pt and prof2 to None.

Reply:
I'd prefer not to default prof_2pt here (I don't think it's so practical, since one would have to instantiate the default class every time - e.g. at every new value of the scale factor).
Otherwise, I've changed the ordering and defaulted prof2